### PR TITLE
Implement Runtime APIs

### PR DESCRIPTION
**Pull request by rphmeier**
This builds on #1401, providing more guide updates and an implementation of all the runtime API primitives and functions (in the runtime). This doesn't provide the implementation of the Runtime API Subsystem, which is filed in #1419

### CODE REVIEW

File: primitives/src/v1.rs [hunk 1]
```diff
@@ -476,3 +479,169 @@ pub struct AvailableData {
...
+		)
+	}
+}
+
+/// Information about a core which is currently occupied.
+#[derive(Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(PartialEq, Debug))]
+pub struct OccupiedCore<N = BlockNumber> {
+	/// The ID of the para occupying the core.
+	pub para: Id,
```
Comment (drahnr) [inline primitives/src/v1.rs:542]: in this context, does para imply both parachain and parathread?

File: roadmap/implementers-guide/src/runtime-api/README.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+Node-side code is also responsible for detecting and reporting misbehavior performed by other validators, and the set of Runtime APIs needs to provide methods for observing live disputes and submitting reports as transactions.
+
+The next sections will contain information on specific runtime APIs.
+
+## Validators
+
+Yields the validator-set at the state of a given block. This validator set is always the one responsible for backing parachains in the child of the provided block.
+
+```rust
+fn validators() -> Vec<ValidatorId>;
```
Comment (drahnr) [inline roadmap/implementers-guide/src/runtime-api/README.md:33]: Should this potentially take an argument, such as the `block_number` or is that implicit?

File: roadmap/implementers-guide/src/types/candidate.md [hunk 3]
```diff
@@ -64,7 +72,7 @@ This struct is pure description of the candidate, in a lightweight format.
 /// A unique descriptor of the candidate receipt.
 struct CandidateDescriptor {
 	/// The ID of the para this is a candidate for.
-	para_id: Id,
+	para_id: ParaId,
```
Comment (drahnr) [inline roadmap/implementers-guide/src/types/candidate.md:75]:
```suggestion
	para: ParaId,
```
consistency

File: roadmap/implementers-guide/src/runtime-api/README.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+Get the session index that is expected at the child of a block.
+
+In the [`Initializer`](../runtime/initializer.md) module, session changes are buffered by one block. The session index of the child of any block is always predictable by that block's state.
+
+This session index can be used to derive a [`SigningContext`](../types/candidate.md#signing-context).
+
+```rust
+/// Returns the session index expected at a child of the block.
+fn session_index_for_child() -> SessionIndex;
```
Comment (drahnr) [inline roadmap/implementers-guide/src/runtime-api/README.md:167]: :+1:

File: primitives/src/v1.rs (hunk 1 above)
Comment (rphmeier) [reply primitives/src/v1.rs:544]: In this module, the type is defined as `Id`, and elsewhere we import as `ParaId`. I would prefer to have a separate PR to rename `struct Id` to `struct ParaId`, as it used to be `primitives::parachain::Id`, but now is `primitives::v1::Id` which is less semantically clear.

File: primitives/src/v1.rs [hunk 5]
```diff
@@ -476,3 +479,199 @@ pub struct AvailableData {
...
+pub enum OccupiedCoreAssumption {
+    /// The candidate occupying the core was made available and included to free the core.
+	#[codec(index = "0")]
+    Included,
+    /// The candidate occupying the core timed out and freed the core without advancing the para.
+	#[codec(index = "1")]
+    TimedOut,
+    /// The core was not occupied to begin with.
+	#[codec(index = "2")]
+    Free,
```
Comment (ordian) [inline primitives/src/v1.rs:606]: nit: mixed tabs and spaces

File: roadmap/implementers-guide/src/types/candidate.md (hunk 3 above)
Comment (rphmeier) [reply roadmap/implementers-guide/src/types/candidate.md]: (have altered for consistency in the other direction)

File: node/core/runtime-api/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "polkadot-node-core-runtime-api"
+version = "0.1.0"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2018"
+
+[dependencies]
+futures = "0.3.5"
+log = "0.4.8"
+sp-api = { git = "https://github.com/paritytech/substrate", branch = "master" }
```
Comment (drahnr) [inline node/core/runtime-api/Cargo.toml:10]: should we consider creating patches/cargo overrides in the workspace?

File: primitives/src/v1.rs [hunk 7]
```diff
@@ -476,3 +479,221 @@ pub struct AvailableData {
...
+}
+
+impl GroupRotationInfo {
+	/// Returns the index of the group needed to validate the core at the given index, assuming
+	/// the given number of cores.
+	///
+	/// `core_index` should be less than `cores`, which in turn should be less than u32::max().
+	pub fn group_for_core(&self, core_index: CoreIndex, cores: usize) -> GroupIndex {
+		if self.group_rotation_frequency == 0 { return GroupIndex(core_index.0) }
+		if cores == 0 { return GroupIndex(0) }
```
Comment (drahnr) [inline primitives/src/v1.rs:502]: is `cores == 0` a valid case?

File: primitives/src/v1.rs [hunk 8]
```diff
@@ -476,3 +479,221 @@ pub struct AvailableData {
...
+
+		let cores = sp_std::cmp::max(cores, u32::max_value() as usize);
```
Comment (drahnr) [inline primitives/src/v1.rs:504]: According to the doc comment this should be `cmp::min(..)`

File: primitives/src/v1.rs [hunk 9]
```diff
@@ -476,3 +479,221 @@ pub struct AvailableData {
...
+	pub group_rotation_frequency: N,
+	/// The current block number.
+	pub now: N,
+}
+
+impl GroupRotationInfo {
+	/// Returns the index of the group needed to validate the core at the given index, assuming
+	/// the given number of cores.
+	///
+	/// `core_index` should be less than `cores`, which in turn should be less than u32::max().
```
Comment (ordian) [inline primitives/src/v1.rs:499]:
```suggestion
	/// `core_index` should be less than `cores`, which in turn should be less than or equal to `u32::max()`.
```
or
```suggestion
	/// `core_index` should be less than `cores`, which is capped at `u32::max()`.
```

File: runtime/parachains/src/runtime_api_impl/v1.rs [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+/// Implementation for the `candidate_pending_availability` function of the runtime API.
+pub fn candidate_pending_availability<T: initializer::Trait>(para_id: ParaId)
+	-> Option<CommittedCandidateReceipt<T::Hash>>
+{
+	<inclusion::Module<T>>::candidate_pending_availability(para_id)
+}
+
+/// Implementatino for the `candidate_events` function of the runtime API.
```
Comment (montekki) [inline runtime/parachains/src/runtime_api_impl/v1.rs:271]:
```suggestion
/// Implementation for the `candidate_events` function of the runtime API.
```

File: primitives/src/v1.rs (hunk 9 above)
Comment (ordian) [reply primitives/src/v1.rs:499]: actually 4 billion cores sounds like too many, I wonder if it should be limited by millions instead

File: node/core/runtime-api/Cargo.toml (hunk 6 above)
Comment (rphmeier) [reply node/core/runtime-api/Cargo.toml:10]: why?

File: primitives/src/v1.rs (hunk 8 above)
Comment (rphmeier) [reply primitives/src/v1.rs:504]: whoops, nice catch

File: primitives/src/v1.rs (hunk 7 above)
Comment (rphmeier) [reply primitives/src/v1.rs:502]: Yeah, if there are no parachains registered and no parathread cores, I think we'd have 0 cores.

File: primitives/src/v1.rs (hunk 9 above)
Comment (rphmeier) [reply primitives/src/v1.rs:499]: this is an arithmetic limit, not a practical one. I expect that we in practice won't have more than a few thousand cores.

Comment (rphmeier) [reply primitives/src/v1.rs:499]: `Vec::<u8>::with_capacity(usize::max())` will fail even though, while legal under API, it is not a practical amount of memory to allocate.

File: node/core/runtime-api/Cargo.toml (hunk 6 above)
Comment (rphmeier) [reply node/core/runtime-api/Cargo.toml:10]: Seems beyond scope of PR. Ideally we'd have this stuff on crates.io

**Pull request merged**